src/Microsoft.Health.Fhir.SqlServer/Features/Schema/Sql/Sprocs/AcquireWatchdogLease.sql

```diff
@@ -0,0 +1,302 @@
+      SET @msg = convert(varchar(1000),'Remove old/forced leases:'+@msg)
+      EXECUTE dbo.LogEvent @Process='AcquireWatchdogLease',@Status='Info',@Mode=@Mode,@Target='WatchdogLeases',@Action='Delete',@Rows=@Rows,@Text=@msg
+    END
+  END
+
+  SET @NumberOfWorkers = 1 + (SELECT count(*) 
+                                FROM (SELECT LeaseHolder FROM dbo.WatchdogLeases WITH (NOLOCK) WHERE LeaseHolder <> @Worker
+                                      UNION 
+                                      SELECT LeaseRequestor FROM dbo.WatchdogLeases WITH (NOLOCK) WHERE LeaseRequestor <> @Worker AND LeaseRequestor <> ''
+                                     ) A
+                             )
+
+  SET @Mode = convert(varchar(100),@Mode+' N='+convert(varchar(10),@NumberOfWorkers))
+
+  -- Prepare the row in table - TABLOCKX
+  IF NOT EXISTS (SELECT * FROM dbo.WatchdogLeases WITH (NOLOCK) WHERE Watchdog = @Watchdog) -- Minimize time when TABLOCKX is needed
+    INSERT INTO dbo.WatchdogLeases (Watchdog, LeaseEndTime, LeaseRequestTime) 
+      SELECT @Watchdog, dateadd(day,-10,@Now), dateadd(day,-10,@Now) 
+        WHERE NOT EXISTS (SELECT * FROM dbo.WatchdogLeases WITH (TABLOCKX) WHERE Watchdog = @Watchdog)
+  
+  SET @LeaseEndTime = dateadd(second,@LeasePeriodSec,@Now)
+  
+  SET @WatchdogNumber = (SELECT count(*) FROM dbo.WatchdogLeases WITH (NOLOCK))
+  --PRINT '@WatchdogNumber = '+convert(varchar,@WatchdogNumber)
+
+  SET @NotLeasedWatchdogNumber = (SELECT count(*) FROM dbo.WatchdogLeases WITH (NOLOCK) WHERE LeaseHolder = '' OR LeaseEndTime < @Now)
+  --PRINT '@NotLeasedWatchdogNumber = '+convert(varchar,@NotLeasedWatchdogNumber)
+  
+  SET @MyLeasesNumber = (SELECT count(*) FROM dbo.WatchdogLeases WITH (NOLOCK) WHERE LeaseHolder = @Worker AND LeaseEndTime > @Now)
+  --PRINT '@MyLeasesNumber = '+convert(varchar,@MyLeasesNumber)
+  
+  SET @OtherValidRequestsOrLeasesNumber = 
+        (SELECT count(*) 
+           FROM dbo.WatchdogLeases WITH (NOLOCK)
+           WHERE LeaseHolder <> @Worker AND LeaseEndTime > @Now 
+              OR LeaseRequestor <> @Worker AND datediff(second,LeaseRequestTime,@Now) < @LeasePeriodSec
+        )
+  --PRINT '@OtherValidRequestsOrLeasesNumber = '+convert(varchar,@OtherValidRequestsOrLeasesNumber)
+  
+  SET @MyValidRequestsOrLeasesNumber = 
+        (SELECT count(*) 
+           FROM dbo.WatchdogLeases WITH (NOLOCK)
+           WHERE LeaseHolder = @Worker AND LeaseEndTime > @Now 
+              OR LeaseRequestor = @Worker AND datediff(second,LeaseRequestTime,@Now) < @LeasePeriodSec
+        )
+  --PRINT '@MyValidRequestsOrLeasesNumber = '+convert(varchar,@MyValidRequestsOrLeasesNumber)
+  
+  SET @DesiredLeasesNumber = ceiling(1.0 * @WatchdogNumber / @NumberOfWorkers)
+  IF @DesiredLeasesNumber = 0 SET @DesiredLeasesNumber = 1
+  IF @DesiredLeasesNumber = 1 AND @OtherValidRequestsOrLeasesNumber = 1 AND @WatchdogNumber = 1 SET @DesiredLeasesNumber = 0
```

Comment:
It is hit when there is a single watchdog with not expired lease in the table for worker X, and worker Y comes with the same watchdog request. This case should be NOOP, so desired number is set to 0 so that below logic does not do anything.